### PR TITLE
Adding support for CredHub as a backend

### PR DESCRIPTION
**Pull request by AlbertoImpl**
Adding support for CredHub as a backend.
Added optional dependency to `spring-credhub-core`

### CODE REVIEW

**Comment by AlbertoImpl** [comment]
Note that `spring-credhub-core` hasn't been released yet, but I wanted to get your feedback before that happened.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		this.credHubOperations = credHubOperations;
+	}
+
+	@Override
+	public Environment findOne(String application, String profilesList, String label) {
+		if (StringUtils.isEmpty(profilesList)) {
+			profilesList = "default";
+		}
+		if (StringUtils.isEmpty(label)) {
+			label = "master";
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java]: Does master make sense here?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		environment.add(new PropertySource("credhub-" + application, properties));
+
+		return environment;
+	}
+
+	private Map<Object, Object> findProperties(String application, String profile, String label) {
+		String path = "/" + application + "/" + profile + "/" + label;
+
+		return credHubOperations.credentials().findByPath(path)
+								.stream()
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java]: Too far indentation

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java (hunk 1 above)
Comment (AlbertoImpl) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java]: I don't think it does make a lot of sense, however, a lot of the repositories, including `Jdbc` are using `master` as a default label.
I am happy to remove the default and just do not concatenate any label if nothing was specified, but for the sake of consistency, I decided to add that there. What do you think?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java (hunk 2 above)
Comment (AlbertoImpl) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CredhubEnvironmentRepository.java]: Fixed, thanks!

**Comment by AlbertoImpl** [comment]
Just added the missing license headers, thanks for the catch @ryanjbaxter!

**Comment by AlbertoImpl** [comment]
Once `spring-credhub-core` gets released. Where should I add the version @spencergibb?

**Comment by spencergibb** [comment]
What version are you asking about?

**Comment by AlbertoImpl** [comment]
At the moment we are building with `spring-credhub-core` with `2.0.0.BUILD-SNAPSHOT` but a `2.0.0.RELEASE` will come out soon.
For the test to run I hardcoded the version here: https://github.com/spring-cloud/spring-cloud-config/pull/1211/files#diff-27baec15a3f2afe27aa44caf3c8e2e4dR53 but there must be a place where I can set it to `2.0.0.RELEASE` outside of that pom. Am I right?

**Comment by spencergibb** [comment]
In the dependencies pom. You can do that now. Any idea when it will be released?

**Comment by scottfrederick** [comment]
@AlbertoImpl This PR can be updated to use `spring-credhub-core` version `2.0.0.RC1` from the Spring milestone repository.

**Comment by AlbertoImpl** [comment]
Just force-pushed the changes with the new version and moved it to `spring-cloud-config-dependencies/pom.xml`.
Let me know if I need to do something else or I did something wrong @spencergibb 

Thanks!

**Comment by spencergibb** [comment]
@scottfrederick just a reminder we will need a spring-credhub release prior to spring cloud Greenwich GA (mid-Jan).

**Pull request merged**